src/engineio/socket.rs

```diff
@@ -79,80 +37,153 @@ pub struct EngineSocket {
     pub connected: Arc<AtomicBool>,
     last_ping: Arc<Mutex<Instant>>,
     last_pong: Arc<Mutex<Instant>>,
-    host_address: Arc<Mutex<Option<String>>>,
-    tls_config: Arc<Option<TlsConnector>>,
-    custom_headers: Arc<Option<HeaderMap>>,
-    connection_data: Arc<RwLock<Option<HandshakeData>>>,
-    engine_io_mode: Arc<AtomicBool>,
+    connection_data: Arc<HandshakePacket>,
 }
 
-impl EngineSocket {
-    /// Creates an instance of `EngineSocket`.
-    pub fn new(
-        engine_io_mode: bool,
-        tls_config: Option<TlsConnector>,
-        opening_headers: Option<HeaderMap>,
-    ) -> Self {
-        let client = match (tls_config.clone(), opening_headers.clone()) {
-            (Some(config), Some(map)) => ClientBuilder::new()
-                .use_preconfigured_tls(config)
-                .default_headers(map)
-                .build()
-                .unwrap(),
-            (Some(config), None) => ClientBuilder::new()
-                .use_preconfigured_tls(config)
-                .build()
-                .unwrap(),
-            (None, Some(map)) => ClientBuilder::new().default_headers(map).build().unwrap(),
-            (None, None) => Client::new(),
-        };
+pub struct SocketBuilder {
+    url: Url,
+    tls_config: Option<TlsConnector>,
+    headers: Option<HeaderMap>,
+}
+
+impl SocketBuilder {
+    pub fn new(url: Url) -> Self {
+        SocketBuilder {
+            url,
+            headers: None,
+            tls_config: None,
+        }
+    }
+
+    pub fn set_tls_config(mut self, tls_config: TlsConnector) -> Self {
+        self.tls_config = Some(tls_config);
+        self
+    }
+
+    pub fn set_headers(mut self, headers: HeaderMap) -> Self {
+        self.headers = Some(headers);
+        self
+    }
+
+    fn _build(&self) -> Result<(HandshakePacket, Url)> {
+        let mut url = self.url.clone();
+        let url = url.query_pairs_mut().append_pair("EIO", "4").finish();
+
+        // No path add engine.io
+        if url.path() == "/" {
+            url.set_path("/engine.io/");
+        }
+
+        // Start with polling transport
+        let transport =
+            PollingTransport::new(url.clone(), self.tls_config.clone(), self.headers.clone());
+
+        let handshake: HandshakePacket = Packet::try_from(transport.poll()?)?.try_into()?;
+
+        // update the base_url with the new sid
+        let url = url
+            .query_pairs_mut()
+            .append_pair("sid", &handshake.sid[..])
+            .finish();
+
+        Ok((handshake, url.clone()))
+    }
+
+    pub fn build(&self) -> Result<Socket<PollingTransport>> {
+        let (handshake, url) = self._build()?;
+
+        // Make a polling transport with new sid
+        let transport = PollingTransport::new(url, self.tls_config.clone(), self.headers.clone());
+
+        // If we can't upgrade or upgrade fails use polling
+        Ok(Socket::new(transport, handshake))
```

Comment:
If build was a consuming method (taking parameter `self`), we could get rid of all those `clone` statements on the member variables.

Reply:
I was only able to get rid of one of the clone statements (as the shared _build function cannot consume self) by making them consume self. Changed the build methods to consume self so we are free to make more improvements in the future.

Reply:
refactored a bit more was able to remove another clone